node/heartbeat/monitorEdgeCases_test.go

```diff
@@ -0,0 +1,483 @@
+package heartbeat_test
+
+import (
+	"testing"
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/node/heartbeat"
+	"github.com/ElrondNetwork/elrond-go/node/heartbeat/storage"
+	"github.com/ElrondNetwork/elrond-go/node/mock"
+	"github.com/stretchr/testify/assert"
+)
+
+var pkValidator = "pk"
+
+func createMonitor(
+	storer heartbeat.HeartbeatStorageHandler,
+	genesisTime time.Time,
+	maxDurationPeerUnresponsive time.Duration,
+	timer heartbeat.Timer,
+) *heartbeat.Monitor {
+	mon, _ := heartbeat.NewMonitor(
+		&mock.MarshalizerFake{},
+		maxDurationPeerUnresponsive,
+		map[uint32][]string{0: {pkValidator}},
+		genesisTime,
+		&mock.MessageHandlerStub{},
+		storer,
+		timer,
+	)
+
+	return mon
+}
+
+const unresponsiveDuration = time.Second * 50
+const tenSeconds = 10
+const twentySeconds = 20
+const hundredSeconds = 100
+const hundredFiftySeconds = 150
+const twoHundredSeconds = 200
+
+// v: |.................................
+// o: |___________|.........|___________
+//    0           20        100   150   200
+func TestMonitor_ObserverGapValidatorOffline(t *testing.T) {
+	t.Parallel()
+
+	storer, _ := storage.NewHeartbeatDbStorer(mock.NewStorerMock(), &mock.MarshalizerFake{})
+	timer := &mock.MockTimer{}
+
+	genesisTime := timer.Now()
+	mon1 := createMonitor(storer, genesisTime, unresponsiveDuration, timer)
+
+	timer.SetSeconds(tenSeconds)
+	heartBeats := mon1.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], false, 0, 10)
+
+	timer.SetSeconds(twentySeconds)
+	heartBeats = mon1.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], false, 0, 20)
+
+	timer.SetSeconds(hundredSeconds)
+	mon2 := createMonitor(storer, genesisTime, unresponsiveDuration, timer)
+
+	timer.SetSeconds(hundredFiftySeconds)
+	heartBeats = mon2.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], false, 0, 50)
+
+	timer.SetSeconds(twoHundredSeconds)
+	heartBeats = mon2.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], false, 0, 100)
+}
+
+// v: |_________________________________
+// o: |___________|.........|___________
+//    0    10    20        100   150   200
+func TestMonitor_ObserverGapValidatorOnline(t *testing.T) {
+	t.Parallel()
+
+	storer, _ := storage.NewHeartbeatDbStorer(mock.NewStorerMock(), &mock.MarshalizerFake{})
+
+	timer := &mock.MockTimer{}
+	genesisTime := timer.Now()
+
+	mon1 := createMonitor(storer, genesisTime, unresponsiveDuration, timer)
+	mon1.AddHeartbeatMessageToMap(&heartbeat.Heartbeat{Pubkey: []byte(pkValidator)})
+
+	heartBeats := mon1.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], true, 0, 0)
+
+	timer.SetSeconds(tenSeconds)
+	//mon1.AddHeartbeatMessageToMap(&heartbeat.Heartbeat{Pubkey: []byte(pkValidator)})
```

Comment:
Shouldn't this line be uncommented? Remove if not.

Reply:
Agree, might be removed because the message has been sent in the unresponsiveDuration window

Reply:
Removed